aten/src/TH/generic/THTensorRandom.cpp

```diff
@@ -444,28 +473,36 @@ void THTensor_(multinomial)(THLongTensor *self, THGenerator *_generator, THTenso
 #endif
 
 #if defined(TH_REAL_IS_BYTE)
-void THTensor_(getRNGState)(THGenerator *_generator, THTensor *self)
+void THTensor_(getRNGState)(at::Generator *_generator, THTensor *self)
 {
-  std::lock_guard<std::mutex> lock(_generator->mutex);
-  static const size_t size = sizeof(THGeneratorState);
-  THGeneratorState *rng_state;
+  // See Note [Thread-safety and Generators]
+  std::lock_guard<std::mutex> lock(_generator->mutex_);
+  static const size_t size = sizeof(at::CPUGenerator);
+  at::CPUGenerator* rng_state;
   THTensor_(resize1d)(self, size);
   THArgCheck(THTensor_(nElement)(self) == size, 1, "RNG state is wrong size");
   THArgCheck(THTensor_(isContiguous)(self), 1, "RNG state needs to be contiguous");
-  rng_state = (THGeneratorState *)self->data<scalar_t>();
-  THGeneratorState_copy(rng_state, &_generator->gen_state);
+  rng_state = (at::CPUGenerator*)self->data<scalar_t>();
+  auto cast_generator = static_cast<at::CPUGenerator*>(_generator);
+  memcpy(rng_state, cast_generator, size);
```

Comment:
OK, I looked into the situation here.

Basically, I think the easiest fix is to create another POD class which contains just the state of the tensor, and then memcpy that in and out (relying on CPUGenerator to have its vtable setup correctly). Even better would be to add some sort of magic header at the top so we can do a consistency check to make sure the user isn't feeding us garbage.

We had a similar problem the last time changes were made to RNG state in #4041, read the comments there, it is essentially the same problem, just with different clothes.

cc @yf225 @colesbury